segmentstore/server/src/main/java/io/pravega/segmentstore/server/SegmentStoreMetrics.java

```diff
@@ -154,36 +168,57 @@ public OperationProcessor(int containerId) {
             this.operationQueueSize = STATS_LOGGER.createStats(MetricsNames.OPERATION_QUEUE_SIZE, this.containerTag);
             this.operationsInFlight = STATS_LOGGER.createStats(MetricsNames.OPERATION_PROCESSOR_IN_FLIGHT, this.containerTag);
             this.operationQueueWaitTime = STATS_LOGGER.createStats(MetricsNames.OPERATION_QUEUE_WAIT_TIME, this.containerTag);
-            this.operationProcessorDelay = STATS_LOGGER.createStats(MetricsNames.OPERATION_PROCESSOR_DELAY_MILLIS, this.containerTag);
             this.operationCommitLatency = STATS_LOGGER.createStats(MetricsNames.OPERATION_COMMIT_LATENCY, this.containerTag);
             this.operationLatency = STATS_LOGGER.createStats(MetricsNames.OPERATION_LATENCY, this.containerTag);
             this.memoryCommitLatency = STATS_LOGGER.createStats(MetricsNames.OPERATION_COMMIT_MEMORY_LATENCY, this.containerTag);
             this.memoryCommitCount = STATS_LOGGER.createStats(MetricsNames.OPERATION_COMMIT_MEMORY_COUNT, this.containerTag);
             this.processOperationsLatency = STATS_LOGGER.createStats(MetricsNames.PROCESS_OPERATIONS_LATENCY, this.containerTag);
             this.processOperationsBatchSize = STATS_LOGGER.createStats(MetricsNames.PROCESS_OPERATIONS_BATCH_SIZE, this.containerTag);
+            this.operationProcessorCacheDelay = STATS_LOGGER.createStats(
```

Comment:
Why do we need three separate metrics here? We can do this using a single metric and change the tag on each publish.

Check out https://github.com/pravega/pravega/blob/master/shared/metrics/src/main/java/io/pravega/shared/metrics/MetricsLogAppender.java

We use a single metric (ok, two of them `MetricsNames.LOG_ERRORS` or `MetricsNames.LOG_WARNINGS`), then in `recordEvent` we use `DYNAMIC_LOGGER` to record a value and attach a tag.

What I think we can do in the throttling metrics is define 3 tags (Cache, DurableDataLog and Batching - or better, just get them out of `ThrottlerCalculator.DelayResult.getThrottlerName().toString()`) and then use a single metric (you can use the one we had before) to report the delay value with the appropriate number.

Reply:
After reading the code, seems in the case there are two tags - container (x1) and throttler (x3) - so it's a best practice to pre-create all possible tag values combinations in advance, and use enum/map to locate them.
Please avoid dynamically generating tag values asap. (Tag values should be, in most cases, constant)

Reply:
The code could be like this:
this.containerTag = containerTag(containerId);
this.throttler-cache = throttlerTag("cache");
this.throttler-batching = throttlerTag("batching");
...
DYNAMIC_LOGGER.reportGaugeValue(MetricsNames.OPERATION_PROCESSOR_DELAY_MILLIS, millis, containerTag, whateverThrottlerTag);

Reply:
I attempted using the `reportGaugeValue` functionality. But I had issues passing in primitive numeric types (Could not change the value of the gauge after the initial argument reference was lost, i.e the method returned). According to the micrometer documentation this is the expected result.

Using a `Counter` or a `MeteredEvent` also don't seem appropriate for the use case.

Reply:
For example, I wrote two simple tests and stuck them in `ThrottlerTests`:

```
    @Test
    public void a() throws Exception {
        DYNAMIC_LOGGER.reportGaugeValue("gauge-key", 10);
        log.debug("gauge-key: {}", MetricRegistryUtils.getGauge("gauge-key").value());
        DYNAMIC_LOGGER.reportGaugeValue("gauge-key", 15);
        log.debug("gauge-key: {}", MetricRegistryUtils.getGauge("gauge-key").value());
    }
    @Test
    public void b() throws Exception {
        log.debug("gauge-key before: {}", MetricRegistryUtils.getGauge("gauge-key").value());
        DYNAMIC_LOGGER.reportGaugeValue("gauge-key", 20);
        log.debug("gauge-key after: {}", MetricRegistryUtils.getGauge("gauge-key").value());
    }
```

Which produced the output:

```
Stats enabled
gauge-key: 10.0
gauge-key: 15.0
Stats enabled
gauge-key before: 15.0
gauge-key after: 15.0
```

Reply:
Please refer to MetricsProviderTest.java for similar test cases, in particular compare how the metrics registry gets initialized.

Reply:
> 
> 
> Please refer to MetricsProviderTest.java for similar test cases, in particular compare how the metrics registry gets initialized.

I'm not sure I understand. I don't see any difference in usages, with the exception of `MetricsProviderTest.java` calling `freezeGaugeValue` before returning.  But that still leaves the question about why the `reportOpenTransacitons` call is able to work properly.

Reply:
Unit tests are supposed to be independent, and be run in any sequence.
Seems the test b() carries the result of test a (15), it is a test leaking. The first line of test b() should receive an nullpointerexception as the gauge wasn't created by that time.
Please refer to MetricsProviderTest again, look for @Before which annotates the method to run before EVERY test case.